{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "f9edc9a6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3.8.8 (default, Apr 13 2021, 19:58:26) \n",
      "[GCC 7.3.0]\n"
     ]
    }
   ],
   "source": [
    "#!export PYTHONPATH=/v/xgrp0j/pnet_prostate_paper:$PYTHONPATH\n",
    "!export PYTHONPATH=~/pnet_prostate_paper:$PYTHONPATH\n",
    "from os.path import dirname, realpath\n",
    "import sys\n",
    "import kaleido\n",
    "print(sys.version)\n",
    "from analysis.vis_utils import get_reactome_pathway_names\n",
    "#from config_path import PATHWAY_PATH\n",
    "from analysis.figure_3.setup import saving_dir\n",
    "\n",
    "current_dir = dirname(realpath(\"__file__\"))\n",
    "module_path = current_dir\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from os.path import join\n",
    "import matplotlib.pyplot as plt\n",
    "import plotly.graph_objects as go\n",
    "from os.path import dirname, realpath\n",
    "\n",
    "from analysis.vis_utils import get_reactome_pathway_names\n",
    "#from config_path import PATHWAY_PATH\n",
    "from analysis.figure_3.setup import saving_dir\n",
    "\n",
    "current_dir = dirname(realpath(\"__file__\"))\n",
    "module_path = current_dir\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from os.path import join\n",
    "import matplotlib.pyplot as plt\n",
    "import plotly.graph_objects as go"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "38ffbe7a",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "'''\n",
    "first layer\n",
    "'''\n",
    "\n",
    "def get_first_layer_df(nlargest):\n",
    "    features_weights = pd.read_csv(join(module_path, './extracted/gradient_importance_0.csv'), index_col=[0, 1])\n",
    "    features_weights['layer'] = 0\n",
    "    nodes_per_layer0 = features_weights[['layer']]\n",
    "    features_weights = features_weights[['coef']]\n",
    "\n",
    "    all_weights = pd.read_csv(join(module_path, './extracted/node_importance_graph_adjusted.csv'), index_col=0)\n",
    "    genes_weights = all_weights[all_weights.layer == 1]\n",
    "    nodes_per_layer1 = genes_weights[['layer']]\n",
    "    genes_weights = genes_weights[['coef_combined']]\n",
    "    # genes_weights = genes_weights[[col_name]]\n",
    "    genes_weights.columns = ['coef']\n",
    "\n",
    "    nodes_per_layer_df = pd.concat([nodes_per_layer0, nodes_per_layer1])\n",
    "    print(genes_weights.head())\n",
    "    print('genes_weights', genes_weights)\n",
    "\n",
    "    node_weights = [features_weights, genes_weights]\n",
    "\n",
    "    df = get_first_layer(node_weights, number_of_best_nodes=nlargest[0], col_name='coef', include_others=True)\n",
    "    # print df.head()\n",
    "    # df.to_csv(join(saving_dir,'first_layer.csv'))\n",
    "    first_layer_df = df[['source', 'target', 'value', 'layer']]\n",
    "    return first_layer_df\n",
    "\n",
    "\n",
    "'''\n",
    "first layer \n",
    "df.head()\n",
    "source\ttarget\tvalue\tdirection\tlayer\n",
    "\n",
    "high_nodes_df\n",
    "index coef\tcoef_combined\tcoef_combined2\tcoef_combined_zscore\tcoef_graph\tlayer\tnode_id\n",
    "\n",
    "important_node_connections_df\n",
    "source\ttarget\tlayer\tvalue\tvalue_abs\tchild_sum_target\tchild_sum_source\tvalue_normalized_by_target\tvalue_normalized_by_source\ttarget_importance\tsource_importance\tA\tB\tvalue_final\tvalue_old\tsource_fan_out\tsource_fan_out_error\ttarget_fan_in\ttarget_fan_in_error\tvalue_final_corrected\n",
    "'''\n",
    "\n",
    "\n",
    "def encode_nodes(df):\n",
    "    source = df['source']\n",
    "    target = df['target']\n",
    "    all_node_labels = list(np.unique(np.concatenate([source, target])))\n",
    "    n_nodes = len(all_node_labels)\n",
    "    node_code = range(n_nodes)\n",
    "    df_encoded = df.replace(all_node_labels, node_code)\n",
    "    return df_encoded, all_node_labels, node_code\n",
    "\n",
    "\n",
    "def get_nlargeest_ind(S):\n",
    "    # ind_source = (S - S.median()).abs() > 3. * S.std()\n",
    "    ind_source = (S - S.median()).abs() > 2. * S.std()\n",
    "    ret = min([10, int(sum(ind_source))])\n",
    "    return ret\n",
    "\n",
    "\n",
    "def get_pathway_names(all_node_ids):\n",
    "    pathways_names = get_reactome_pathway_names()\n",
    "    all_node_labels = pd.Series(all_node_ids).replace(list(pathways_names['id']), list(pathways_names['name']))\n",
    "    return all_node_labels\n",
    "\n",
    "\n",
    "def get_nodes_per_layer_filtered(nodes_per_layer_df, all_node_ids, all_node_labels):\n",
    "    all_node_ids_df = pd.DataFrame(index=all_node_ids)\n",
    "    nodes_per_layer_filtered_df = nodes_per_layer_df.join(all_node_ids_df, how='right')\n",
    "    nodes_per_layer_filtered_df = nodes_per_layer_filtered_df.fillna(0)\n",
    "    nodes_per_layer_filtered_df = nodes_per_layer_filtered_df.groupby(nodes_per_layer_filtered_df.index).min()\n",
    "    mapping_dict = dict(zip(all_node_ids, all_node_labels))\n",
    "    nodes_per_layer_filtered_df.index = nodes_per_layer_filtered_df.index.map(lambda x: mapping_dict[x])\n",
    "    mapping_dict = {y: x for x, y in all_node_labels.to_dict().iteritems()}\n",
    "    nodes_per_layer_filtered_df.index = nodes_per_layer_filtered_df.index.map(lambda x: mapping_dict[x])\n",
    "    return nodes_per_layer_filtered_df\n",
    "\n",
    "\n",
    "# features_weights = pd.read_csv(join(module_path,'./extracted/gradient_importance_0.csv'), index_col =[0,1])\n",
    "# features_weights = features_weights.reset_index()\n",
    "# features_weights.columns= ['target', 'source', 'value']\n",
    "# features_weights['layer'] = 0\n",
    "# features_weights.head()\n",
    "\n",
    "def get_links_with_first_layer():\n",
    "    '''\n",
    "        :return: all_links_df: dataframe with all the connections in the model (with first layer)\n",
    "        '''\n",
    "    links = []\n",
    "\n",
    "    link = pd.read_csv(join(module_path, './extracted/gradient_importance_0.csv'), index_col=[0, 1])\n",
    "    link = link.reset_index()\n",
    "    link.columns = ['target', 'source', 'value']\n",
    "    link['layer'] = 0\n",
    "    links.append(link)\n",
    "    for l in range(1, 7):\n",
    "        link = pd.read_csv(join(module_path, './extracted/link_weights_{}.csv'.format(l)), index_col=0)\n",
    "        link.index.name = 'source'\n",
    "        link = link.reset_index()\n",
    "        link_unpivoted = pd.melt(link, id_vars=['source'], var_name='target', value_name='value')\n",
    "        link_unpivoted['layer'] = l\n",
    "        link_unpivoted = link_unpivoted[link_unpivoted.value != 0.]\n",
    "        link_unpivoted = link_unpivoted.drop_duplicates(subset=['source', 'target'])\n",
    "        links.append(link_unpivoted)\n",
    "    all_links_df = pd.concat(links, axis=0, sort=True)\n",
    "\n",
    "    return all_links_df\n",
    "\n",
    "\n",
    "def get_links():\n",
    "    '''\n",
    "    :return: all_links_df: dataframe with all the connections in the model (except first layer)\n",
    "    '''\n",
    "    links = []\n",
    "    for l in range(1, 7):\n",
    "        link = pd.read_csv(join(module_path, './extracted/link_weights_{}.csv'.format(l)), index_col=0)\n",
    "        link.index.name = 'source'\n",
    "        link = link.reset_index()\n",
    "        link_unpivoted = pd.melt(link, id_vars=['source'], var_name='target', value_name='value')\n",
    "        link_unpivoted['layer'] = l\n",
    "        link_unpivoted = link_unpivoted[link_unpivoted.value != 0.]\n",
    "        link_unpivoted = link_unpivoted.drop_duplicates(subset=['source', 'target'])\n",
    "        links.append(link_unpivoted)\n",
    "    all_links_df = pd.concat(links, axis=0)\n",
    "    return all_links_df\n",
    "\n",
    "\n",
    "def get_high_nodes(node_importance, nlargest, column):\n",
    "    '''\n",
    "    get n largest nodes in each layer\n",
    "    :param:  node_importance: dataframe with coef_combined  and layer columns\n",
    "    :return: list of high node names\n",
    "    '''\n",
    "    layers = np.sort(node_importance.layer.unique())\n",
    "    high_nodes = []\n",
    "    for i, l in enumerate(layers):\n",
    "        if type(nlargest) == list:\n",
    "            n = nlargest[i]\n",
    "        else:\n",
    "            n = nlargest\n",
    "        high = list(node_importance[node_importance.layer == l].nlargest(n, columns=column).index)\n",
    "        high_nodes.extend(high)\n",
    "    return high_nodes\n",
    "\n",
    "\n",
    "def filter_nodes(node_importance, high_nodes, add_others=True):\n",
    "    high_nodes_df = node_importance[node_importance.index.isin(high_nodes)].copy()\n",
    "    # add others:\n",
    "\n",
    "    if add_others:\n",
    "        layers = list(node_importance.layer.unique())\n",
    "        names = ['others{}'.format(l) for l in layers]\n",
    "        names = names + ['root']\n",
    "        layers.append(np.max(layers) + 1)\n",
    "        print(layers)\n",
    "        print(names)\n",
    "        data = {'index': names, 'layer': layers}\n",
    "        df = pd.DataFrame.from_dict(data)\n",
    "        df = df.set_index('index')\n",
    "        high_nodes_df = high_nodes_df.append(df)\n",
    "    return high_nodes_df\n",
    "\n",
    "\n",
    "def filter_connections(df, high_nodes, add_unk=False):\n",
    "    \"\"\"\n",
    "    :param df: dataframe [ source, target, layer]\n",
    "    :param high_nodes: list of high node ids\n",
    "    :param add_unk: boolean , add other connections or not\n",
    "    :return: ret, dataframe [ source, target, layer] with filtered connection based on the high_nodes\n",
    "    \"\"\"\n",
    "    high_nodes.append('root')\n",
    "\n",
    "    def apply_others(row):\n",
    "        # print row\n",
    "        if not row['source'] in high_nodes_list:\n",
    "            row['source'] = 'others' + str(row['layer'])\n",
    "\n",
    "        if not row['target'] in high_nodes_list:\n",
    "            row['target'] = 'others' + str(row['layer'] + 1)\n",
    "        return row\n",
    "\n",
    "    layers_id = np.sort(df.layer.unique())\n",
    "    high_nodes_list = high_nodes\n",
    "    layer_dfs = []\n",
    "    for i, l in enumerate(layers_id):\n",
    "        layer_df = df[df.layer == l].copy()\n",
    "        ind1 = layer_df.source.isin(high_nodes_list)\n",
    "        ind2 = layer_df.target.isin(high_nodes_list)\n",
    "        if add_unk:\n",
    "            layer_df = layer_df[ind1 | ind2]\n",
    "            layer_df = layer_df.apply(apply_others, axis=1)\n",
    "\n",
    "        else:\n",
    "            layer_df = layer_df[ind1 & ind2]\n",
    "\n",
    "        layer_df = layer_df.groupby(['source', 'target']).agg({'value': 'sum', 'layer': 'min'})\n",
    "\n",
    "        layer_dfs.append(layer_df)\n",
    "    ret = pd.concat(layer_dfs)\n",
    "    return ret\n",
    "\n",
    "\n",
    "def get_x_y(df_encoded, layers_nodes):\n",
    "    '''\n",
    "    :param df_encoded: datafrme with columns (source  target  value) representing the network\n",
    "    :param layers_nodes: data frame with index (nodes ) and one columns (layer) representing the layer of the node\n",
    "    :return: x, y positions onf each node\n",
    "    '''\n",
    "\n",
    "    # node_id = range(len(layers_nodes))\n",
    "    # node_weights = pd.DataFrame([node_id, layers_nodes], columns=['node_id', 'node_name'])\n",
    "    # print node_weights\n",
    "    # node weight is the max(sum of input edges, sum of output edges)\n",
    "\n",
    "    def rescale(val, in_min, in_max, out_min, out_max):\n",
    "        print(val, in_min, in_max, out_min, out_max)\n",
    "        if in_min == in_max:\n",
    "            return val\n",
    "        return out_min + (val - in_min) * ((out_max - out_min) / (in_max - in_min))\n",
    "\n",
    "    source_weights = df_encoded.groupby(by='source')['value'].sum()\n",
    "    target_weights = df_encoded.groupby(by='target')['value'].sum()\n",
    "\n",
    "    node_weights = pd.concat([source_weights, target_weights])\n",
    "    node_weights = node_weights.to_frame()\n",
    "    node_weights = node_weights.groupby(node_weights.index).max()\n",
    "\n",
    "    node_weights = node_weights.join(layers_nodes)\n",
    "    # node_weights['value'] = node_weights.groupby('layer')['value'].apply(lambda x: rescale(x, min(x), max(x), 0., 1.))\n",
    "\n",
    "    ind = node_weights.index.str.contains('others')\n",
    "\n",
    "    others_value = node_weights.loc[ind, 'value']\n",
    "    print('others_value', others_value)\n",
    "    node_weights.loc[ind, 'value'] = 0.\n",
    "    node_weights.sort_values(by=['layer', 'value'], ascending=False, inplace=True)\n",
    "    print('others_value', others_value)\n",
    "    node_weights.loc[others_value.index, 'value'] = others_value\n",
    "    n_layers = len(layers_nodes['layer'].unique())\n",
    "\n",
    "    node_weights['x'] = (node_weights['layer'] - 2) * 0.1 + 0.16\n",
    "    # node_weights['x'] = (node_weights['layer']-2) *0.15 + 0.16\n",
    "    ind = node_weights.layer == 0\n",
    "    node_weights.loc[ind, 'x'] = 0.01\n",
    "    ind = node_weights.layer == 1\n",
    "    node_weights.loc[ind, 'x'] = 0.08\n",
    "    ind = node_weights.layer == 2\n",
    "    node_weights.loc[ind, 'x'] = 0.14\n",
    "\n",
    "    xs = np.linspace(0.14, 1, 6, endpoint=False)\n",
    "    for i, x in enumerate(xs[1:]):\n",
    "        print(i, x)\n",
    "        ind = node_weights.layer == i + 3\n",
    "        node_weights.loc[ind, 'x'] = x\n",
    "\n",
    "    # node_weights.loc[ind,'x' ] = 0.3\n",
    "    # ind = node_weights.layer==4\n",
    "    # node_weights.loc[ind,'x' ] = 0.4\n",
    "    # ind = node_weights.layer==5\n",
    "    # node_weights.loc[ind,'x' ] = 0.5\n",
    "    # ind = node_weights.layer==6\n",
    "    # node_weights.loc[ind,'x' ] = 0.6\n",
    "    # ind = node_weights.layer==7\n",
    "    # node_weights.loc[ind,'x' ] = 0.7\n",
    "\n",
    "    print('node_weights', node_weights)\n",
    "    # node_weights.to_csv('node_weights.csv')\n",
    "    dd = node_weights.groupby('layer')['value'].transform(pd.Series.sum)\n",
    "    node_weights['layer_weight'] = dd\n",
    "    node_weights['y'] = node_weights.groupby('layer')['value'].transform(pd.Series.cumsum)\n",
    "    node_weights['y'] = (node_weights['y'] - .5 * node_weights['value']) / (1.5 * node_weights['layer_weight'])\n",
    "    # node_weights['y'] = (node_weights['y'] -  .5 * node_weights['value']) /(node_weights['layer_weight'])\n",
    "\n",
    "    # root node\n",
    "    # ind = node_weights.layer==7\n",
    "    # node_weights.loc[ind,'x' ] = 0.67\n",
    "    node_weights.loc[ind, 'y'] = 0.33\n",
    "\n",
    "    print('node_weights', node_weights['x'], node_weights['y'])\n",
    "    node_weights.sort_index(inplace=True)\n",
    "    node_weights.to_csv('node_weights_original.csv')\n",
    "    # node_weights.to_csv('xy.csv')\n",
    "    return node_weights['x'], node_weights['y']\n",
    "\n",
    "\n",
    "# def get_data_trace(links, all_node_labels, node_pos, layers, node_colors=None):\n",
    "def get_data_trace(links, node_df, height, width, fontsize=6):\n",
    "    '''\n",
    "    linkes: dataframe [source, target, value]\n",
    "    all_node_labels: list of real node names\n",
    "    node_pos: tuples of (x, y) for neach node\n",
    "    layers: dataframe with index is node name and columns is layer\n",
    "    node_colors, list of colors for each node (Same order as all_node_labels )\n",
    "    '''\n",
    "\n",
    "    # df = pd.DataFrame(node_colors, index=all_node_labels,columns=['color'] )\n",
    "    # df.to_csv('all_node_labels.csv')\n",
    "\n",
    "    all_node_labels = node_df.short_name.values\n",
    "    node_colors = node_df.color.values\n",
    "    x = node_df.x.values\n",
    "    y = node_df.y.values\n",
    "\n",
    "    def rescale(val, in_min, in_max, out_min, out_max):\n",
    "        print(val, in_min, in_max, out_min, out_max)\n",
    "        if in_min == in_max:\n",
    "            return val\n",
    "        return out_min + (val - in_min) * ((out_max - out_min) / (in_max - in_min))\n",
    "\n",
    "    x = rescale(x, min(x), max(x), 0.01, .98)\n",
    "    # y = rescale(y, min(y), max(y), 0.01, .96)\n",
    "\n",
    "    data_trace = dict(\n",
    "        type='sankey',\n",
    "        arrangement='snap',\n",
    "        domain=dict(\n",
    "            x=[0, 1.],\n",
    "            y=[0, 1.]\n",
    "        ),\n",
    "        orientation=\"h\",\n",
    "        valueformat=\".0f\",\n",
    "        node=dict(\n",
    "            pad=2,\n",
    "            thickness=10,\n",
    "            line=dict(\n",
    "                color=\"white\",\n",
    "                width=.5\n",
    "            ),\n",
    "            label=all_node_labels,\n",
    "            x=x,\n",
    "            y=y,\n",
    "            color=node_colors,\n",
    "\n",
    "        ),\n",
    "        link=dict(\n",
    "            source=links['source'],\n",
    "            target=links['target'],\n",
    "            value=links['value'],\n",
    "            color=links['color']\n",
    "        )\n",
    "    )\n",
    "\n",
    "    layout = dict(\n",
    "        height=height,\n",
    "        width=width,\n",
    "        margin=go.layout.Margin(\n",
    "            l=0,  # left margin\n",
    "            r=0,  # right margin\n",
    "            b=0.1,  # bottom margin\n",
    "            t=8,  # top margin\n",
    "        ),\n",
    "        font=dict(\n",
    "            size=fontsize, family='Arial',\n",
    "        )\n",
    "    )\n",
    "    return data_trace, layout\n",
    "\n",
    "\n",
    "def get_node_colors(all_node_labels, remove_others=True):\n",
    "    def color_to_hex(color):\n",
    "        r, g, b, a = [255 * c for c in color]\n",
    "        c = '#%02X%02X%02X' % (r, g, b)\n",
    "        return c\n",
    "\n",
    "    color_idx = np.linspace(1, 0, len(all_node_labels))\n",
    "    cmp = plt.cm.Reds\n",
    "    node_colors = {}\n",
    "    for i, node in zip(color_idx, all_node_labels):\n",
    "        if 'other' in node:\n",
    "            if remove_others:\n",
    "                c = (255, 255, 255, 0.0)\n",
    "            else:\n",
    "                c = (232, 232, 232, 0.5)\n",
    "        else:\n",
    "            colors = list(cmp(i))\n",
    "            colors = [int(255 * c) for c in colors]\n",
    "            colors[-1] = 0.7  # set alpha\n",
    "            c = colors\n",
    "        node_colors[node] = c\n",
    "\n",
    "    return node_colors\n",
    "\n",
    "\n",
    "def get_edge_colors(df, node_colors_dict, remove_others=True):\n",
    "    colors = []\n",
    "    for i, row in df.iterrows():\n",
    "        if ('others' in row['source']) or ('others' in row['target']):\n",
    "            if remove_others:\n",
    "                colors.append('rgba(255, 255, 255, 0.0)')\n",
    "            else:\n",
    "                colors.append('rgba(192, 192, 192, 0.2)')\n",
    "        else:\n",
    "            #             colors.append('rgba(252, 186, 3, .4)')\n",
    "            base_color = [l for l in node_colors_dict[row['source']]]\n",
    "            base_color[-1] = 0.2\n",
    "            base_color = 'rgba{}'.format(tuple(base_color))\n",
    "            # colors.append('rgba(192,192,192,0.5)')\n",
    "            colors.append(base_color)\n",
    "    return colors\n",
    "\n",
    "\n",
    "def get_node_colors_ordered(high_nodes_df, col_name, remove_others=True):\n",
    "    '''\n",
    "    high_nodes_df: data frame with index (node names) andcolumns [layer, coef_combined]\n",
    "    '''\n",
    "    node_colors = {}\n",
    "    layers = high_nodes_df.layer.unique()\n",
    "    for l in layers:\n",
    "        nodes_ordered = high_nodes_df[high_nodes_df.layer == l].sort_values(col_name, ascending=False).index\n",
    "        node_colors.update(get_node_colors(nodes_ordered, remove_others))\n",
    "\n",
    "    return node_colors\n",
    "\n",
    "\n",
    "def get_first_layer(node_weights, number_of_best_nodes, col_name='coef', include_others=True):\n",
    "    gene_weights = node_weights[1].copy()\n",
    "    feature_weights = node_weights[0].copy()\n",
    "\n",
    "    gene_weights = gene_weights[[col_name]]\n",
    "    feature_weights = feature_weights[[col_name]]\n",
    "\n",
    "    if number_of_best_nodes == 'auto':\n",
    "        S = gene_weights[col_name].sort_values()\n",
    "        n = get_nlargeest_ind(S)\n",
    "        top_genes = list(gene_weights.nlargest(n, col_name).index)\n",
    "    else:\n",
    "        top_genes = list(gene_weights.nlargest(number_of_best_nodes, col_name).index)\n",
    "\n",
    "    # gene normalization\n",
    "    print('top_genes', top_genes)\n",
    "    genes = gene_weights.loc[top_genes]\n",
    "    genes[col_name] = np.log(1. + genes[col_name].abs())\n",
    "\n",
    "    print(genes.shape)\n",
    "    print(genes.head())\n",
    "\n",
    "    df = feature_weights\n",
    "\n",
    "    if include_others:\n",
    "        df = df.reset_index()\n",
    "\n",
    "        df.columns = ['target', 'source', 'value']\n",
    "        df['target'] = df['target'].map(lambda x: x if x in top_genes else 'others1')\n",
    "        df = df.groupby(['source', 'target']).sum()\n",
    "\n",
    "        df = df.reset_index()\n",
    "        df.columns = ['source', 'target', 'value']\n",
    "\n",
    "        print('df groupby')\n",
    "        print(df)\n",
    "\n",
    "    else:\n",
    "        df = feature_weights.loc[top_genes]\n",
    "        df = df.reset_index()\n",
    "        df.columns = ['target', 'source', 'value']\n",
    "\n",
    "    df['direction'] = df['value'] >= 0.\n",
    "    df['value'] = abs(df['value'])\n",
    "\n",
    "    df['source'] = df['source'].replace('mut_important', 'mutation')\n",
    "    df['source'] = df['source'].replace('cnv', 'copy number')\n",
    "    df['source'] = df['source'].replace('cnv_amp', 'amplification')\n",
    "    df['source'] = df['source'].replace('cnv_del', 'deletion')\n",
    "    df['source'] = df['source'].replace('gene_expression', 'methylation')\n",
    "    df['layer'] = 0\n",
    "\n",
    "    # normalize features by gene\n",
    "    # groups = df.groupby('target')\n",
    "\n",
    "    # sum1 = groups['value'].transform(np.sum)\n",
    "    df['value'] = df['value'] / df.groupby('target')['value'].transform(np.sum)\n",
    "    df = df[df.value > 0.0]\n",
    "\n",
    "    # multiply by the gene importance\n",
    "    # genes\n",
    "    df = pd.merge(df, genes, left_on='target', right_index=True, how='left')\n",
    "    print(df.shape)\n",
    "    df.coef.fillna(10.0, inplace=True)\n",
    "    df.value = df.value * df.coef * 150.\n",
    "\n",
    "    print(df.shape)\n",
    "\n",
    "    return df\n",
    "\n",
    "\n",
    "def get_fromated_network(links, high_nodes_df, col_name, remove_others):\n",
    "    # get node colors\n",
    "    node_colors_dict = get_node_colors_ordered(high_nodes_df, col_name, remove_others)\n",
    "\n",
    "    print('node_colors_dict', node_colors_dict)\n",
    "    # exception for first layer\n",
    "    node_colors_dict['amplification'] = (224, 123, 57, 0.7)  # amps\n",
    "    node_colors_dict['deletion'] = (1, 55, 148, 0.7)  # deletion\n",
    "    node_colors_dict['mutation'] = (105, 189, 210, 0.7)  # mutation\n",
    "    node_colors_dict['methylation'] = (100, 55, 200, 0.7)  # mutation\n",
    "    node_colors_dict['expression_tpm'] = (60, 55, 50, 0.7)  # mutation\n",
    "\n",
    "    # get colors\n",
    "    links['color'] = get_edge_colors(links, node_colors_dict, remove_others)\n",
    "\n",
    "    # get node colors\n",
    "    for key, value in node_colors_dict.items():\n",
    "        node_colors_dict[key] = 'rgba{}'.format(tuple(value))\n",
    "\n",
    "    # remove links with no values\n",
    "    links = links.dropna(subset=['value'], axis=0)\n",
    "\n",
    "    # enocde nodes (convert node names into numbers)\n",
    "    linkes_filtred_encoded_df, all_node_labels, node_code = encode_nodes(links)\n",
    "\n",
    "    # get node per layers\n",
    "    node_layers_df = high_nodes_df[['layer']]\n",
    "\n",
    "    # remove self connection\n",
    "    ind = linkes_filtred_encoded_df.source == linkes_filtred_encoded_df.target\n",
    "    linkes_filtred_encoded_df = linkes_filtred_encoded_df[~ind]\n",
    "\n",
    "    # make sure we positive values for all edges\n",
    "    linkes_filtred_encoded_df.value = linkes_filtred_encoded_df.value.abs()\n",
    "\n",
    "    # linkes_filtred_encoded_df = linkes_filtred_encoded_df.fillna(0.001)\n",
    "\n",
    "    # linkes_filtred_encoded_df = linkes_filtred_encoded_df[~linkes_filtred_encoded_df['value'].isna()]\n",
    "\n",
    "    x, y = get_x_y(links, node_layers_df)\n",
    "\n",
    "    # shorten names\n",
    "\n",
    "    def get_short_names(all_node_labels):\n",
    "        PATHWAY_PATH = \"/v//projects/methyldeeplearn-pallag/paurel/pnet_aging/_database/pathways/\"\n",
    "        df = pd.read_excel(join(PATHWAY_PATH, 'pathways_short_names.xlsx'), index_col=0)\n",
    "        mapping_dict = {}\n",
    "        for k, v in zip(df['Full name'].values, df['Short name (Eli)'].values):\n",
    "            mapping_dict[k] = str(v)\n",
    "\n",
    "        all_node_labels_short = []\n",
    "        for l in all_node_labels:\n",
    "            short_name = l\n",
    "            if l in mapping_dict.keys() and not mapping_dict[l] == 'nan':\n",
    "                short_name = mapping_dict[l]\n",
    "\n",
    "            if 'others' in short_name:\n",
    "                short_name = 'residual'\n",
    "            if 'root' in short_name:\n",
    "                short_name = 'outcome'\n",
    "\n",
    "            all_node_labels_short.append(short_name)\n",
    "\n",
    "        # all_node_labels_short = []\n",
    "        # for n in all_node_labels:\n",
    "        #     to_be_added = str(n)\n",
    "        #     if len(to_be_added) >=30:\n",
    "        #         to_be_added = to_be_added[:25] + ' ...'\n",
    "        #     if 'others' in to_be_added :\n",
    "        #         to_be_added ='residual'\n",
    "        #     if 'root' in to_be_added :\n",
    "        #         to_be_added ='outcome'\n",
    "        #         # to_be_added ='root'\n",
    "        #     # to_be_added = '{}({},{})'.format(to_be_added,str(round(x[n],2)),str(round(y[n],2)) )\n",
    "        #     all_node_labels_short.append(to_be_added)\n",
    "        return all_node_labels_short\n",
    "\n",
    "    # node_colors_list = []\n",
    "    # for l in all_node_labels:\n",
    "    #     node_colors_list.append(node_colors_dict[l])\n",
    "    # xy_pos = (x, y)\n",
    "    # node_colors_df = pd.DataFrame(node_colors_dict)\n",
    "\n",
    "    node_colors_list = []\n",
    "    for l in all_node_labels:\n",
    "        node_colors_list.append(node_colors_dict[l])\n",
    "\n",
    "    all_node_labels_short = get_short_names(all_node_labels)\n",
    "\n",
    "    data = np.column_stack((node_code, node_colors_list, all_node_labels_short))\n",
    "    nodes_df = pd.DataFrame(data, columns=['code', 'color', 'short_name'], index=all_node_labels)\n",
    "    nodes_df = nodes_df.join(x, how='left')\n",
    "    nodes_df = nodes_df.join(y, how='left')\n",
    "\n",
    "    print(nodes_df.head())\n",
    "    print(nodes_df.shape)\n",
    "    # return linkes_filtred_encoded_df, all_node_labels_short, xy_pos, node_layers_df, node_colors_list\n",
    "    return linkes_filtred_encoded_df, nodes_df\n",
    "\n",
    "\n",
    "# #remove self connections\n",
    "\n",
    "\n",
    "def get_MDM4_nodes(links_df):\n",
    "    import networkx as nx\n",
    "    net = nx.from_pandas_edgelist(links_df, 'target', 'source', create_using=nx.DiGraph())\n",
    "    net.name = 'reactome'\n",
    "    # add root node\n",
    "    roots = [n for n, d in net.in_degree() if d == 0]\n",
    "    root_node = 'root'\n",
    "    edges = [(root_node, n) for n in roots]\n",
    "    net.add_edges_from(edges)\n",
    "    # convert to tree\n",
    "    tree = nx.bfs_tree(net, 'root')\n",
    "    traces = list(nx.all_simple_paths(tree, 'root', 'HMGA1_cg01745499'))\n",
    "    #traces = list(nx.all_simple_paths(tree, 'root', 'MDM4'))\n",
    "    print(len(traces))\n",
    "    all_odes = []\n",
    "    for t in traces:\n",
    "        print ('-->'.join(t))\n",
    "        all_odes.extend(t)\n",
    "\n",
    "    nodes = np.unique(all_odes)\n",
    "    return nodes\n",
    "    # MDM4_subnet = tree.subgraph(nodes)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "9b1ff74f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/v/projects/methyldeeplearn-pallag/paurel/pnet_aging/_database/pathways/Reactome/\n",
      "high_nodes ['NCOR2_cg22580512', 'RPL31_cg22809047', 'VGF_cg04084157', 'SSRP1_cg01511567', 'NHLRC1_cg22736354', 'Late Phase of HIV Life Cycle', 'Nuclear Events (kinase and transcription factor activation)', 'FGFR2 alternative splicing', 'InlA-mediated entry of Listeria monocytogenes into host cells', 'Influenza Viral RNA Transcription and Replication', 'HIV Life Cycle', 'Signaling by NTRK1 (TRKA)', 'Influenza Life Cycle', 'Signaling by FGFR2', 'Host Interactions of HIV factors', 'HIV Infection', 'Listeria monocytogenes entry into host cells', 'Signaling by NTRKs', 'Signaling by NOTCH1', 'Influenza Infection', 'Metabolism of amino acids and derivatives', 'Signaling by Receptor Tyrosine Kinases', 'Diseases of signal transduction', 'Metabolism of vitamins and cofactors', 'Signaling by NOTCH', 'Neuronal System', 'Immune System', 'Metabolism', 'Transport of small molecules', 'Metabolism of RNA']\n",
      "[2, 1, 4, 5, 3, 6, 7]\n",
      "['others2', 'others1', 'others4', 'others5', 'others3', 'others6', 'root']\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "<ipython-input-2-7641e233ce22>:174: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  high_nodes_df = high_nodes_df.append(df)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0\n",
      "mdm4_nodes []\n",
      "                      coef\n",
      "A1BG_cg22568540   4.718241\n",
      "A1CF_cg03817621   0.290398\n",
      "A2M_cg12058490    0.203810\n",
      "A4GNT_cg17687282  0.000000\n",
      "AAAS_cg00559473   0.000000\n",
      "genes_weights                            coef\n",
      "A1BG_cg22568540    4.718241e+00\n",
      "A1CF_cg03817621    2.903982e-01\n",
      "A2M_cg12058490     2.038102e-01\n",
      "A4GNT_cg17687282   0.000000e+00\n",
      "AAAS_cg00559473    0.000000e+00\n",
      "...                         ...\n",
      "ZWILCH_cg18864581  1.008189e+00\n",
      "ZWINT_cg18534872   7.060559e-01\n",
      "ZWINT_cg21206959   7.402805e-01\n",
      "ZZZ3_cg04127303    2.578373e-08\n",
      "ZZZ3_cg24762437    5.413097e-15\n",
      "\n",
      "[15591 rows x 1 columns]\n",
      "top_genes ['NCOR2_cg22580512', 'RPL31_cg22809047', 'VGF_cg04084157', 'SSRP1_cg01511567', 'NHLRC1_cg22736354']\n",
      "(5, 1)\n",
      "                       coef\n",
      "NCOR2_cg22580512   6.049994\n",
      "RPL31_cg22809047   6.014208\n",
      "VGF_cg04084157     5.937945\n",
      "SSRP1_cg01511567   5.694583\n",
      "NHLRC1_cg22736354  5.685286\n",
      "df groupby\n",
      "            source             target         value\n",
      "0  gene_expression   NCOR2_cg22580512    423.110400\n",
      "1  gene_expression  NHLRC1_cg22736354    293.502230\n",
      "2  gene_expression   RPL31_cg22809047    408.201720\n",
      "3  gene_expression   SSRP1_cg01511567    296.252750\n",
      "4  gene_expression     VGF_cg04084157    378.155100\n",
      "5  gene_expression            others1  32598.441262\n",
      "(6, 6)\n",
      "(6, 6)\n",
      "node_colors_dict {'Metabolism of amino acids and derivatives': [103, 0, 12, 0.7], 'Signaling by Receptor Tyrosine Kinases': [187, 20, 25, 0.7], 'Diseases of signal transduction': [241, 68, 50, 0.7], 'Metabolism of vitamins and cofactors': [251, 138, 106, 0.7], 'Signaling by NOTCH': [252, 201, 180, 0.7], 'others5': (232, 232, 232, 0.5), 'Late Phase of HIV Life Cycle': [103, 0, 12, 0.7], 'Nuclear Events (kinase and transcription factor activation)': [187, 20, 25, 0.7], 'FGFR2 alternative splicing': [241, 68, 50, 0.7], 'InlA-mediated entry of Listeria monocytogenes into host cells': [251, 138, 106, 0.7], 'Influenza Viral RNA Transcription and Replication': [252, 201, 180, 0.7], 'others2': (232, 232, 232, 0.5), 'HIV Infection': [103, 0, 12, 0.7], 'Listeria monocytogenes entry into host cells': [187, 20, 25, 0.7], 'Signaling by NTRKs': [241, 68, 50, 0.7], 'Signaling by NOTCH1': [251, 138, 106, 0.7], 'Influenza Infection': [252, 201, 180, 0.7], 'others4': (232, 232, 232, 0.5), 'HIV Life Cycle': [103, 0, 12, 0.7], 'Signaling by NTRK1 (TRKA)': [187, 20, 25, 0.7], 'Influenza Life Cycle': [241, 68, 50, 0.7], 'Signaling by FGFR2': [251, 138, 106, 0.7], 'Host Interactions of HIV factors': [252, 201, 180, 0.7], 'others3': (232, 232, 232, 0.5), 'Neuronal System': [103, 0, 12, 0.7], 'Immune System': [187, 20, 25, 0.7], 'Metabolism': [241, 68, 50, 0.7], 'Transport of small molecules': [251, 138, 106, 0.7], 'Metabolism of RNA': [252, 201, 180, 0.7], 'others6': (232, 232, 232, 0.5), 'NCOR2_cg22580512': [103, 0, 12, 0.7], 'RPL31_cg22809047': [187, 20, 25, 0.7], 'VGF_cg04084157': [241, 68, 50, 0.7], 'SSRP1_cg01511567': [251, 138, 106, 0.7], 'NHLRC1_cg22736354': [252, 201, 180, 0.7], 'others1': (232, 232, 232, 0.5), 'root': [103, 0, 12, 0.7], 'mutation': [103, 0, 12, 0.7], 'amplification': [227, 47, 39, 0.7], 'deletion': [252, 159, 129, 0.7], 'methylation': [255, 245, 240, 0.7]}\n",
      "others_value others1    1500.000000\n",
      "others2     506.227700\n",
      "others3     325.965202\n",
      "others4     740.643947\n",
      "others5    1058.781331\n",
      "others6     576.114728\n",
      "Name: value, dtype: float64\n",
      "others_value others1    1500.000000\n",
      "others2     506.227700\n",
      "others3     325.965202\n",
      "others4     740.643947\n",
      "others5    1058.781331\n",
      "others6     576.114728\n",
      "Name: value, dtype: float64\n",
      "0 0.2833333333333333\n",
      "1 0.4266666666666667\n",
      "2 0.5700000000000001\n",
      "3 0.7133333333333334\n",
      "4 0.8566666666666667\n",
      "node_weights                                                           value  layer  \\\n",
      "root                                                1252.997616      7   \n",
      "Neuronal System                                      269.177419      6   \n",
      "Immune System                                        253.725416      6   \n",
      "Metabolism                                           249.810724      6   \n",
      "Transport of small molecules                         244.261997      6   \n",
      "Metabolism of RNA                                    236.022061      6   \n",
      "others6                                              576.114728      6   \n",
      "Metabolism of amino acids and derivatives            215.503181      5   \n",
      "Signaling by Receptor Tyrosine Kinases               207.453485      5   \n",
      "Diseases of signal transduction                      199.407735      5   \n",
      "Metabolism of vitamins and cofactors                 179.269904      5   \n",
      "Signaling by NOTCH                                   169.253507      5   \n",
      "others5                                             1058.781331      5   \n",
      "HIV Infection                                        216.348253      4   \n",
      "Listeria monocytogenes entry into host cells         199.399501      4   \n",
      "Signaling by NTRKs                                   168.464390      4   \n",
      "Signaling by NOTCH1                                  156.532129      4   \n",
      "Influenza Infection                                  156.451986      4   \n",
      "others4                                              740.643947      4   \n",
      "HIV Life Cycle                                       286.153497      3   \n",
      "Signaling by NTRK1 (TRKA)                            222.932298      3   \n",
      "Influenza Life Cycle                                 196.249123      3   \n",
      "Signaling by FGFR2                                   188.905497      3   \n",
      "Host Interactions of HIV factors                     175.440682      3   \n",
      "others3                                              325.965202      3   \n",
      "Late Phase of HIV Life Cycle                         187.679194      2   \n",
      "Nuclear Events (kinase and transcription factor...   149.636165      2   \n",
      "FGFR2 alternative splicing                           117.121301      2   \n",
      "InlA-mediated entry of Listeria monocytogenes i...   115.248438      2   \n",
      "Influenza Viral RNA Transcription and Replication    109.016739      2   \n",
      "others2                                              506.227700      2   \n",
      "NCOR2_cg22580512                                     907.499070      1   \n",
      "RPL31_cg22809047                                     902.131236      1   \n",
      "VGF_cg04084157                                       890.691803      1   \n",
      "SSRP1_cg01511567                                     854.187418      1   \n",
      "NHLRC1_cg22736354                                    852.792971      1   \n",
      "others1                                             1500.000000      1   \n",
      "methylation                                         5907.302498      0   \n",
      "\n",
      "                                                           x  \n",
      "root                                                0.856667  \n",
      "Neuronal System                                     0.713333  \n",
      "Immune System                                       0.713333  \n",
      "Metabolism                                          0.713333  \n",
      "Transport of small molecules                        0.713333  \n",
      "Metabolism of RNA                                   0.713333  \n",
      "others6                                             0.713333  \n",
      "Metabolism of amino acids and derivatives           0.570000  \n",
      "Signaling by Receptor Tyrosine Kinases              0.570000  \n",
      "Diseases of signal transduction                     0.570000  \n",
      "Metabolism of vitamins and cofactors                0.570000  \n",
      "Signaling by NOTCH                                  0.570000  \n",
      "others5                                             0.570000  \n",
      "HIV Infection                                       0.426667  \n",
      "Listeria monocytogenes entry into host cells        0.426667  \n",
      "Signaling by NTRKs                                  0.426667  \n",
      "Signaling by NOTCH1                                 0.426667  \n",
      "Influenza Infection                                 0.426667  \n",
      "others4                                             0.426667  \n",
      "HIV Life Cycle                                      0.283333  \n",
      "Signaling by NTRK1 (TRKA)                           0.283333  \n",
      "Influenza Life Cycle                                0.283333  \n",
      "Signaling by FGFR2                                  0.283333  \n",
      "Host Interactions of HIV factors                    0.283333  \n",
      "others3                                             0.283333  \n",
      "Late Phase of HIV Life Cycle                        0.140000  \n",
      "Nuclear Events (kinase and transcription factor...  0.140000  \n",
      "FGFR2 alternative splicing                          0.140000  \n",
      "InlA-mediated entry of Listeria monocytogenes i...  0.140000  \n",
      "Influenza Viral RNA Transcription and Replication   0.140000  \n",
      "others2                                             0.140000  \n",
      "NCOR2_cg22580512                                    0.080000  \n",
      "RPL31_cg22809047                                    0.080000  \n",
      "VGF_cg04084157                                      0.080000  \n",
      "SSRP1_cg01511567                                    0.080000  \n",
      "NHLRC1_cg22736354                                   0.080000  \n",
      "others1                                             0.080000  \n",
      "methylation                                         0.010000  \n",
      "node_weights root                                                             0.856667\n",
      "Neuronal System                                                  0.713333\n",
      "Immune System                                                    0.713333\n",
      "Metabolism                                                       0.713333\n",
      "Transport of small molecules                                     0.713333\n",
      "Metabolism of RNA                                                0.713333\n",
      "others6                                                          0.713333\n",
      "Metabolism of amino acids and derivatives                        0.570000\n",
      "Signaling by Receptor Tyrosine Kinases                           0.570000\n",
      "Diseases of signal transduction                                  0.570000\n",
      "Metabolism of vitamins and cofactors                             0.570000\n",
      "Signaling by NOTCH                                               0.570000\n",
      "others5                                                          0.570000\n",
      "HIV Infection                                                    0.426667\n",
      "Listeria monocytogenes entry into host cells                     0.426667\n",
      "Signaling by NTRKs                                               0.426667\n",
      "Signaling by NOTCH1                                              0.426667\n",
      "Influenza Infection                                              0.426667\n",
      "others4                                                          0.426667\n",
      "HIV Life Cycle                                                   0.283333\n",
      "Signaling by NTRK1 (TRKA)                                        0.283333\n",
      "Influenza Life Cycle                                             0.283333\n",
      "Signaling by FGFR2                                               0.283333\n",
      "Host Interactions of HIV factors                                 0.283333\n",
      "others3                                                          0.283333\n",
      "Late Phase of HIV Life Cycle                                     0.140000\n",
      "Nuclear Events (kinase and transcription factor activation)      0.140000\n",
      "FGFR2 alternative splicing                                       0.140000\n",
      "InlA-mediated entry of Listeria monocytogenes into host cells    0.140000\n",
      "Influenza Viral RNA Transcription and Replication                0.140000\n",
      "others2                                                          0.140000\n",
      "NCOR2_cg22580512                                                 0.080000\n",
      "RPL31_cg22809047                                                 0.080000\n",
      "VGF_cg04084157                                                   0.080000\n",
      "SSRP1_cg01511567                                                 0.080000\n",
      "NHLRC1_cg22736354                                                0.080000\n",
      "others1                                                          0.080000\n",
      "methylation                                                      0.010000\n",
      "Name: x, dtype: float64 root                                                             0.330000\n",
      "Neuronal System                                                  0.049054\n",
      "Immune System                                                    0.144347\n",
      "Metabolism                                                       0.236110\n",
      "Transport of small molecules                                     0.326149\n",
      "Metabolism of RNA                                                0.413675\n",
      "others6                                                          0.561677\n",
      "Metabolism of amino acids and derivatives                        0.035392\n",
      "Signaling by Receptor Tyrosine Kinases                           0.104855\n",
      "Diseases of signal transduction                                  0.171673\n",
      "Metabolism of vitamins and cofactors                             0.233864\n",
      "Signaling by NOTCH                                               0.291102\n",
      "others5                                                          0.492783\n",
      "HIV Infection                                                    0.044031\n",
      "Listeria monocytogenes entry into host cells                     0.128644\n",
      "Signaling by NTRKs                                               0.203512\n",
      "Signaling by NOTCH1                                              0.269655\n",
      "Influenza Infection                                              0.333354\n",
      "others4                                                          0.515931\n",
      "HIV Life Cycle                                                   0.068344\n",
      "Signaling by NTRK1 (TRKA)                                        0.189933\n",
      "Influenza Life Cycle                                             0.290050\n",
      "Signaling by FGFR2                                               0.382039\n",
      "Host Interactions of HIV factors                                 0.469059\n",
      "others3                                                          0.588814\n",
      "Late Phase of HIV Life Cycle                                     0.052796\n",
      "Nuclear Events (kinase and transcription factor activation)      0.147687\n",
      "FGFR2 alternative splicing                                       0.222728\n",
      "InlA-mediated entry of Listeria monocytogenes into host cells    0.288096\n",
      "Influenza Viral RNA Transcription and Replication                0.351185\n",
      "others2                                                          0.524259\n",
      "NCOR2_cg22580512                                                 0.051208\n",
      "RPL31_cg22809047                                                 0.153320\n",
      "VGF_cg04084157                                                   0.254485\n",
      "SSRP1_cg01511567                                                 0.352943\n",
      "NHLRC1_cg22736354                                                0.449264\n",
      "others1                                                          0.582026\n",
      "methylation                                                      0.333333\n",
      "Name: y, dtype: float64\n",
      "                                 code                     color  \\\n",
      "Diseases of signal transduction     0    rgba(241, 68, 50, 0.7)   \n",
      "FGFR2 alternative splicing          1    rgba(241, 68, 50, 0.7)   \n",
      "HIV Infection                       2     rgba(103, 0, 12, 0.7)   \n",
      "HIV Life Cycle                      3     rgba(103, 0, 12, 0.7)   \n",
      "Host Interactions of HIV factors    4  rgba(252, 201, 180, 0.7)   \n",
      "\n",
      "                                                        short_name         x  \\\n",
      "Diseases of signal transduction    Diseases of signal transduction  0.570000   \n",
      "FGFR2 alternative splicing              FGFR2 alternative splicing  0.140000   \n",
      "HIV Infection                                        HIV Infection  0.426667   \n",
      "HIV Life Cycle                                      HIV Life Cycle  0.283333   \n",
      "Host Interactions of HIV factors  Host Interactions of HIV factors  0.283333   \n",
      "\n",
      "                                         y  \n",
      "Diseases of signal transduction   0.171673  \n",
      "FGFR2 alternative splicing        0.222728  \n",
      "HIV Infection                     0.044031  \n",
      "HIV Life Cycle                    0.068344  \n",
      "Host Interactions of HIV factors  0.469059  \n",
      "(38, 5)\n",
      "[0.57       0.14       0.42666667 0.28333333 0.28333333 0.71333333\n",
      " 0.42666667 0.28333333 0.14       0.14       0.14       0.42666667\n",
      " 0.71333333 0.71333333 0.57       0.57       0.08       0.08\n",
      " 0.71333333 0.14       0.08       0.08       0.28333333 0.57\n",
      " 0.42666667 0.28333333 0.42666667 0.57       0.71333333 0.08\n",
      " 0.01       0.08       0.14       0.28333333 0.42666667 0.57\n",
      " 0.71333333 0.85666667] 0.01 0.8566666666666667 0.01 0.98\n",
      "[0.57       0.14       0.42666667 0.28333333 0.28333333 0.71333333\n",
      " 0.42666667 0.28333333 0.14       0.14       0.14       0.42666667\n",
      " 0.71333333 0.71333333 0.57       0.57       0.08       0.08\n",
      " 0.71333333 0.14       0.08       0.08       0.28333333 0.57\n",
      " 0.42666667 0.28333333 0.42666667 0.57       0.71333333 0.08\n",
      " 0.01       0.08       0.14       0.28333333 0.42666667 0.57\n",
      " 0.71333333 0.85666667] 0.01 0.8566666666666667 0.01 0.98\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'~/_plots/figure3/sankey_youngblood.html'"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "# get reactome pathway ids and names\n",
    "#REACTOM_PATHWAY_PATH = \"/_database/pathways/Reactome/\"\n",
    "reactome_pathway_df = get_reactome_pathway_names()\n",
    "id_to_name_dict = dict(zip(reactome_pathway_df.id, reactome_pathway_df.name))\n",
    "name_to_id_dict = dict(zip(reactome_pathway_df.name, reactome_pathway_df.id))\n",
    "\n",
    "# nlargest= [10, 8, 8, 8, 7, 6]\n",
    "nlargest = [5, 5, 5, 5, 5, 5, 5]\n",
    "# nlargest= 10\n",
    "\n",
    "node_importance = pd.read_csv(join(module_path, './extracted/node_importance_graph_adjusted.csv'), index_col=0)\n",
    "#####\n",
    "# node_importance.coef_combined = node_importance.coef\n",
    "\n",
    "node_id = []\n",
    "for x in node_importance.index:\n",
    "    if x in name_to_id_dict.keys():\n",
    "        node_id.append(name_to_id_dict[x])\n",
    "    else:\n",
    "        node_id.append(x)\n",
    "node_importance['node_id'] = node_id\n",
    "\n",
    "col_name = 'coef'\n",
    "\n",
    "first_layer_nodes = node_importance[node_importance.layer == 1].copy()\n",
    "other_layer_nodes = node_importance[node_importance.layer != 1].copy()\n",
    "high_nodes_first_layer = get_high_nodes(first_layer_nodes, nlargest=nlargest, column='coef_combined')\n",
    "# high_nodes_first_layer = get_high_nodes(first_layer_nodes, nlargest=nlargest, column='coef')\n",
    "high_nodes_pathways = get_high_nodes(other_layer_nodes, nlargest=nlargest, column='coef')\n",
    "# high_nodes_pathways= get_high_nodes(other_layer_nodes, nlargest=nlargest, column='coef_combined')\n",
    "high_nodes = high_nodes_first_layer + high_nodes_pathways\n",
    "print('high_nodes', high_nodes)\n",
    "high_nodes_df = filter_nodes(node_importance, high_nodes)\n",
    "\n",
    "high_nodes_ids = list(high_nodes_df.node_id.values)\n",
    "\n",
    "links_df = get_links()\n",
    "# links_df = get_links_with_first_layer()\n",
    "\n",
    "'''\n",
    "MDM4\n",
    "'''\n",
    "mdm4_nodes = get_MDM4_nodes(links_df)\n",
    "mdm4_nodes_names = []\n",
    "for n in mdm4_nodes:\n",
    "    if n in id_to_name_dict.keys():\n",
    "        mdm4_nodes_names.append(id_to_name_dict[n])\n",
    "    else:\n",
    "        mdm4_nodes_names.append(n)\n",
    "\n",
    "print('mdm4_nodes', mdm4_nodes_names)\n",
    "\n",
    "ind = links_df.source == links_df.target\n",
    "links_df = links_df[~ind]\n",
    "\n",
    "# # keep important nodes only\n",
    "links_df = filter_connections(links_df, high_nodes_ids, add_unk=True)\n",
    "\n",
    "links_df = links_df.reset_index()\n",
    "\n",
    "# print links_df.head()\n",
    "links_df['value_abs'] = links_df.value.abs()\n",
    "\n",
    "links_df['child_sum_target'] = links_df.groupby('target').value_abs.transform(np.sum)\n",
    "links_df['child_sum_source'] = links_df.groupby('source').value_abs.transform(np.sum)\n",
    "links_df['value_normalized_by_target'] = 100 * links_df.value_abs / links_df.child_sum_target\n",
    "links_df['value_normalized_by_source'] = 100 * links_df.value_abs / links_df.child_sum_source\n",
    "\n",
    "#\n",
    "node_importance['coef_combined_normalized_by_layer'] = 100. * node_importance[col_name] / \\\n",
    "                                                       node_importance.groupby('layer')[col_name].transform(np.sum)\n",
    "\n",
    "node_importance_ = node_importance[['node_id', 'coef_combined_normalized_by_layer', col_name]].copy()\n",
    "#\n",
    "#\n",
    "node_importance_['coef_combined_normalized_by_layer'] = np.log(\n",
    "    1. + node_importance_.coef_combined_normalized_by_layer)\n",
    "node_importance_normalized = node_importance_[['node_id', 'coef_combined_normalized_by_layer']]\n",
    "node_importance_normalized = node_importance_normalized.set_index('node_id')\n",
    "node_importance_normalized.columns = ['target_importance']\n",
    "#\n",
    "links_df_ = pd.merge(links_df, node_importance_normalized, left_on='target', right_index=True, how='left')\n",
    "node_importance_normalized.columns = ['source_importance']\n",
    "links_df_ = pd.merge(links_df_, node_importance_normalized, left_on='source', right_index=True, how='left')\n",
    "\n",
    "#\n",
    "def adjust_values(links_df_in):\n",
    "    df = links_df_in.copy()\n",
    "    df['A'] = df.value_normalized_by_source * df.source_importance\n",
    "    df['B'] = df.value_normalized_by_target * df.target_importance\n",
    "    df['value_final'] = df[[\"A\", \"B\"]].min(axis=1)\n",
    "    #\n",
    "    df['value_old'] = df.value\n",
    "    df.value = df.value_final\n",
    "    #\n",
    "    df['source_fan_out'] = df.groupby('source').value_final.transform(np.sum)\n",
    "    df['source_fan_out_error'] = np.abs(df.source_fan_out - 100. * df.source_importance)\n",
    "\n",
    "    df['target_fan_in'] = df.groupby('target').value_final.transform(np.sum)\n",
    "    df['target_fan_in_error'] = np.abs(df.target_fan_in - 100. * df.target_importance)\n",
    "    #\n",
    "    #\n",
    "    ind = df.source.str.contains('others')\n",
    "    df['value_final_corrected'] = df.value_final\n",
    "    df.loc[ind, 'value_final_corrected'] = df[ind].value_final + df[ind].target_fan_in_error\n",
    "    ind = df.target.str.contains('others')\n",
    "\n",
    "    df.loc[ind, 'value_final_corrected'] = df[ind].value_final_corrected + df[ind].source_fan_out_error\n",
    "\n",
    "    df.value = df.value_final_corrected\n",
    "    return df\n",
    "\n",
    "df = adjust_values(links_df_)\n",
    "# df.to_csv('links_df.csv')\n",
    "important_node_connections_df = df.replace(id_to_name_dict)\n",
    "\n",
    "# important_node_connections_df.to_csv('important_node_connections_df.csv')\n",
    "# high_nodes_df.to_csv('high_nodes_df.csv')\n",
    "\n",
    "high_nodes_df = high_nodes_df[[col_name, 'layer']]\n",
    "\n",
    "# add feature nodes\n",
    "high_nodes_df.loc['mutation'] = [1, 0]\n",
    "high_nodes_df.loc['amplification'] = [1, 0]\n",
    "high_nodes_df.loc['deletion'] = [1, 0]\n",
    "high_nodes_df.loc['methylation'] = [1, 0]\n",
    "# high_nodes_df.loc['other1'] = [1, 1]\n",
    "# high_nodes_df.loc['hidden'] = [1, 8]\n",
    "\n",
    "# add first layer\n",
    "first_layer_df = get_first_layer_df(nlargest)\n",
    "links_df = pd.concat([first_layer_df, important_node_connections_df], sort=True).reset_index()\n",
    "\n",
    "\n",
    "linkes_filtred_, nodes_df = get_fromated_network(links_df, high_nodes_df, col_name=col_name, remove_others=False)\n",
    "\n",
    "scale = 1.\n",
    "width = 600. / scale\n",
    "height = 0.5 * width / scale\n",
    "# linkes_filtred_.to_csv('linkes_filtred.csv')\n",
    "# nodes_df.to_csv('nodes_df.csv')\n",
    "data_trace, layout = get_data_trace(linkes_filtred_, nodes_df, height, width)\n",
    "fig = dict(data=[data_trace], layout=layout)\n",
    "fig = go.Figure(fig)\n",
    "filename = join(saving_dir, 'sankey_print.pdf')\n",
    "fig.write_image(filename, scale=1, width=width, height=height, format='pdf')\n",
    "\n",
    "filename = join(saving_dir, 'sankey_print.png')\n",
    "fig.write_image(filename, scale=5, width=width, height=height, format='png')\n",
    "\n",
    "from plotly.offline import plot\n",
    "scale = 0.5\n",
    "width = 600. / scale\n",
    "height = 0.5 * width\n",
    "data_trace, layout = get_data_trace(linkes_filtred_, nodes_df, height, width, fontsize=12)\n",
    "fig = dict(data=[data_trace], layout=layout)\n",
    "filename = 'sankey_youngblood.html'\n",
    "filename = join(saving_dir, filename)\n",
    "plot(fig, filename=filename)\n",
    "\n",
    "\n",
    "#\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "d62afbbb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'data': [{'type': 'sankey',\n",
       "   'arrangement': 'snap',\n",
       "   'domain': {'x': [0, 1.0], 'y': [0, 1.0]},\n",
       "   'orientation': 'h',\n",
       "   'valueformat': '.0f',\n",
       "   'node': {'pad': 2,\n",
       "    'thickness': 10,\n",
       "    'line': {'color': 'white', 'width': 0.5},\n",
       "    'label': array(['Diseases of signal transduction', 'FGFR2 alternative splicing',\n",
       "           'HIV Infection', 'HIV Life Cycle',\n",
       "           'Host Interactions of HIV factors', 'Immune System',\n",
       "           'Influenza Infection', 'Influenza Life Cycle',\n",
       "           'Influenza Viral RNA Transcription and Replication',\n",
       "           'InlA-mediated entry of Listeria monocytogenes into host cells',\n",
       "           'Late Phase of HIV Life Cycle',\n",
       "           'Listeria monocytogenes entry into host cells', 'Metabolism',\n",
       "           'Metabolism of RNA', 'Metabolism of amino acids and derivatives',\n",
       "           'Metabolism of vitamins and cofactors', 'NCOR2_cg22580512',\n",
       "           'NHLRC1_cg22736354', 'Neuronal System',\n",
       "           'Nuclear Events (kinase and transcription factor activation)',\n",
       "           'RPL31_cg22809047', 'SSRP1_cg01511567', 'Signaling by FGFR2',\n",
       "           'Signaling by NOTCH', 'Signaling by NOTCH1',\n",
       "           'Signaling by NTRK1 (TRKA)', 'Signaling by NTRKs',\n",
       "           'Signaling by Receptor Tyrosine Kinases',\n",
       "           'Transport of small molecules', 'VGF_cg04084157', 'methylation',\n",
       "           'residual', 'residual', 'residual', 'residual', 'residual',\n",
       "           'residual', 'outcome'], dtype=object),\n",
       "    'x': array([0.6515748 , 0.15893701, 0.4873622 , 0.32314961, 0.32314961,\n",
       "           0.8157874 , 0.4873622 , 0.32314961, 0.15893701, 0.15893701,\n",
       "           0.15893701, 0.4873622 , 0.8157874 , 0.8157874 , 0.6515748 ,\n",
       "           0.6515748 , 0.09019685, 0.09019685, 0.8157874 , 0.15893701,\n",
       "           0.09019685, 0.09019685, 0.32314961, 0.6515748 , 0.4873622 ,\n",
       "           0.32314961, 0.4873622 , 0.6515748 , 0.8157874 , 0.09019685,\n",
       "           0.01      , 0.09019685, 0.15893701, 0.32314961, 0.4873622 ,\n",
       "           0.6515748 , 0.8157874 , 0.98      ]),\n",
       "    'y': array([0.17167348, 0.2227283 , 0.04403121, 0.06834432, 0.46905903,\n",
       "           0.14434693, 0.33335375, 0.29004978, 0.35118458, 0.28809639,\n",
       "           0.05279616, 0.12864422, 0.23611023, 0.41367482, 0.03539217,\n",
       "           0.23386385, 0.05120775, 0.44926383, 0.04905429, 0.14768657,\n",
       "           0.15332037, 0.35294348, 0.38203933, 0.29110198, 0.26965521,\n",
       "           0.18993334, 0.2035119 , 0.1048545 , 0.32614893, 0.25448459,\n",
       "           0.33333333, 0.58202567, 0.52425941, 0.5888138 , 0.51593077,\n",
       "           0.49278261, 0.56167681, 0.33      ]),\n",
       "    'color': array(['rgba(241, 68, 50, 0.7)', 'rgba(241, 68, 50, 0.7)',\n",
       "           'rgba(103, 0, 12, 0.7)', 'rgba(103, 0, 12, 0.7)',\n",
       "           'rgba(252, 201, 180, 0.7)', 'rgba(187, 20, 25, 0.7)',\n",
       "           'rgba(252, 201, 180, 0.7)', 'rgba(241, 68, 50, 0.7)',\n",
       "           'rgba(252, 201, 180, 0.7)', 'rgba(251, 138, 106, 0.7)',\n",
       "           'rgba(103, 0, 12, 0.7)', 'rgba(187, 20, 25, 0.7)',\n",
       "           'rgba(241, 68, 50, 0.7)', 'rgba(252, 201, 180, 0.7)',\n",
       "           'rgba(103, 0, 12, 0.7)', 'rgba(251, 138, 106, 0.7)',\n",
       "           'rgba(103, 0, 12, 0.7)', 'rgba(252, 201, 180, 0.7)',\n",
       "           'rgba(103, 0, 12, 0.7)', 'rgba(187, 20, 25, 0.7)',\n",
       "           'rgba(187, 20, 25, 0.7)', 'rgba(251, 138, 106, 0.7)',\n",
       "           'rgba(251, 138, 106, 0.7)', 'rgba(252, 201, 180, 0.7)',\n",
       "           'rgba(251, 138, 106, 0.7)', 'rgba(187, 20, 25, 0.7)',\n",
       "           'rgba(241, 68, 50, 0.7)', 'rgba(187, 20, 25, 0.7)',\n",
       "           'rgba(251, 138, 106, 0.7)', 'rgba(241, 68, 50, 0.7)',\n",
       "           'rgba(100, 55, 200, 0.7)', 'rgba(232, 232, 232, 0.5)',\n",
       "           'rgba(232, 232, 232, 0.5)', 'rgba(232, 232, 232, 0.5)',\n",
       "           'rgba(232, 232, 232, 0.5)', 'rgba(232, 232, 232, 0.5)',\n",
       "           'rgba(232, 232, 232, 0.5)', 'rgba(103, 0, 12, 0.7)'], dtype=object)},\n",
       "   'link': {'source': 0     30\n",
       "    1     30\n",
       "    2     30\n",
       "    3     30\n",
       "    4     30\n",
       "          ..\n",
       "    58    18\n",
       "    59    12\n",
       "    60     5\n",
       "    61    28\n",
       "    62    13\n",
       "    Name: source, Length: 63, dtype: int64,\n",
       "    'target': 0     16\n",
       "    1     17\n",
       "    2     20\n",
       "    3     21\n",
       "    4     29\n",
       "          ..\n",
       "    58    37\n",
       "    59    37\n",
       "    60    37\n",
       "    61    37\n",
       "    62    37\n",
       "    Name: target, Length: 63, dtype: int64,\n",
       "    'value': 0     907.499070\n",
       "    1     852.792971\n",
       "    2     902.131236\n",
       "    3     854.187418\n",
       "    4     890.691803\n",
       "             ...    \n",
       "    58    269.177419\n",
       "    59    249.810724\n",
       "    60    253.725416\n",
       "    61    244.261997\n",
       "    62    236.022061\n",
       "    Name: value, Length: 63, dtype: float64,\n",
       "    'color': 0      rgba(100, 55, 200, 0.2)\n",
       "    1      rgba(100, 55, 200, 0.2)\n",
       "    2      rgba(100, 55, 200, 0.2)\n",
       "    3      rgba(100, 55, 200, 0.2)\n",
       "    4      rgba(100, 55, 200, 0.2)\n",
       "                    ...           \n",
       "    58       rgba(103, 0, 12, 0.2)\n",
       "    59      rgba(241, 68, 50, 0.2)\n",
       "    60      rgba(187, 20, 25, 0.2)\n",
       "    61    rgba(251, 138, 106, 0.2)\n",
       "    62    rgba(252, 201, 180, 0.2)\n",
       "    Name: color, Length: 63, dtype: object}}],\n",
       " 'layout': {'height': 600.0,\n",
       "  'width': 1200.0,\n",
       "  'margin': layout.Margin({\n",
       "      'b': 0.1, 'l': 0, 'r': 0, 't': 8\n",
       "  }),\n",
       "  'font': {'size': 12, 'family': 'Arial'}}}"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fig"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "585be7c4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'temp-plot.html'"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "plot(fig)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "d827498b",
   "metadata": {},
   "outputs": [],
   "source": [
    "from dash import Dash, html, dcc\n",
    "import plotly.express as px\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "de69c030",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Dash is running on http://127.0.0.1:8052/\n",
      "\n",
      " * Serving Flask app \"__main__\" (lazy loading)\n",
      " * Environment: production\n",
      "\u001b[31m   WARNING: This is a development server. Do not use it in a production deployment.\u001b[0m\n",
      "\u001b[2m   Use a production WSGI server instead.\u001b[0m\n",
      " * Debug mode: on\n"
     ]
    }
   ],
   "source": [
    "# Run this app with `python app.py` and\n",
    "# visit http://127.0.0.1:8050/ in your web browser.\n",
    "\n",
    "\n",
    "\n",
    "app = Dash(__name__)\n",
    "\n",
    "# assume you have a \"long-form\" data frame\n",
    "# see https://plotly.com/python/px-arguments/ for more options\n",
    "df = pd.DataFrame({\n",
    "    \"Fruit\": [\"Apples\", \"Oranges\", \"Bananas\", \"Apples\", \"Oranges\", \"Bananas\"],\n",
    "    \"Amount\": [4, 1, 2, 2, 4, 5],\n",
    "    \"City\": [\"SF\", \"SF\", \"SF\", \"Montreal\", \"Montreal\", \"Montreal\"]\n",
    "})\n",
    "\n",
    "fig = px.bar(df, x=\"Fruit\", y=\"Amount\", color=\"City\", barmode=\"group\")\n",
    "\n",
    "app.layout = html.Div(children=[\n",
    "    html.H1(children='Hello Dash'),\n",
    "\n",
    "    html.Div(children='''\n",
    "        Dash: A web application framework for your data.\n",
    "    '''),\n",
    "\n",
    "    dcc.Graph(\n",
    "        id='example-graph',\n",
    "        figure=fig\n",
    "    )\n",
    "])\n",
    "\n",
    "if __name__ == '__main__':\n",
    "    #app.run_server(debug=True)\n",
    "    app.run_server(debug=True, use_reloader=False,port=8052)  <---- Here"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ea67ad92",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d9a386f7",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "970184b7",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
